code/datums/components/multitile.dm

```diff
@@ -16,45 +16,64 @@
  */
 
 //distance_from_center does not include src itself
-/datum/component/multitile/Initialize(distance_from_center, new_filler_map)
-	if(!distance_from_center || !length(new_filler_map))
+/datum/component/multitile/Initialize(new_filler_map)
+	if(!length(new_filler_map))
 		return COMPONENT_INCOMPATIBLE
 
 	if(!isatom(parent))
 		return COMPONENT_INCOMPATIBLE
+	var/max_height = length(new_filler_map)
+	var/max_width = length(new_filler_map[1]) //it should have the same length on every row
+
+	var/offset_x = 0
+	var/offset_y = 0
 
 	var/atom/owner = parent
-	if(owner.x + distance_from_center > world.maxx || owner.x - distance_from_center < 1)
+
+	for(var/i in 1 to length(new_filler_map))
+		if(length(new_filler_map[i] != max_width))
+			stack_trace("A multitile component was passed a list wich did not have the same length every row. Atom parent is: [parent]")
+			var/obj/machinery/machine = parent
+			machine.deconstruct()
+			return COMPONENT_INCOMPATIBLE
+
+		for(var/j in 1 to length(new_filler_map[i]))
+			if(new_filler_map[i][j] == 2)
+				offset_x = j - ((length(new_filler_map[i]) + 1) / 2)
+				offset_y = i - ((length(new_filler_map) + 1) / 2)
+
+	var/distance_from_center_x = (max_width - 1) / 2
+	var/distance_from_center_y = (max_height - 1) / 2
+
+	if(owner.x - offset_x + distance_from_center_x > world.maxx || owner.x + offset_x - distance_from_center_x < 1)
 		var/obj/machinery/machine = parent
 		machine.deconstruct()
 		return COMPONENT_INCOMPATIBLE
 
-	if(owner.y + distance_from_center > world.maxy || owner.y - distance_from_center < 1)
+	if(owner.y + offset_y + distance_from_center_y > world.maxy || owner.y - offset_y - distance_from_center_y < 1)
 		var/obj/machinery/machine = parent
 		machine.deconstruct()
 		return COMPONENT_INCOMPATIBLE
 
-	var/max_height = length(new_filler_map)
-	var/max_width = length(new_filler_map[1]) //it should have the same length on every row
-	for(var/i in 2 to length(new_filler_map))
-		var/length = length(new_filler_map[i])
-		if(length != max_width)
-			stack_trace("A multitile component was passed a list wich did not have the same length every row. Atom parent is: [parent]")
-			var/obj/machinery/machine = parent
-			machine.deconstruct()
-			return COMPONENT_INCOMPATIBLE
-
 	var/current_height = 0
-	var/current_width = 0
+	var/current_width = 1
+	var/tile_index = 1
 
-	for(var/turf/filler_turf as anything in RANGE_TURFS(distance_from_center, owner))
-		if(new_filler_map[max_height - current_height][max_width - current_width]) // Because the `block()` proc always works from the bottom left to the top right, we have to loop through our nested lists in reverse
+	for(var/turf/filler_turf as anything in block( \
+	owner.x - offset_x - distance_from_center_x, owner.y + offset_y - distance_from_center_y, owner.z, \
+	owner.x - offset_x + distance_from_center_x, owner.y + offset_y + distance_from_center_y, owner.z, \
+	))
```

Comment:
```suggestion
	for(var/turf/filler_turf as anything in block( 
	owner.x - offset_x - distance_from_center_x, owner.y + offset_y - distance_from_center_y, owner.z, 
	owner.x - offset_x + distance_from_center_x, owner.y + offset_y + distance_from_center_y, owner.z, 
	))
```

Reply:
Does `block()` work nicely with multiline arguments?

Reply:
Tested, and it seems to work fine